### ISSUE TITLE
Implementing PRONG

### ISSUE THREAD

**Issue by sergeyf** [fchollet/keras#563]
[This work](http://arxiv.org/pdf/1507.00210v1.pdf) reports very good results with using vanilla SGD when each layer is whitened as part of the optimization procedure. Sort of a fancier (and reportedly more effective) version of batch normalization. How hard would this be to implement in Keras? Algorithm 1 on page 4 looks like it might work within the layers API. Do I have that right?

**Comment by EderSantana** [comment]
@sergeyf indeed this should be interesting. Now that layers support individual `updates` we could just add the proper whitening to the layers who support it. But did I understand it well, they don't do whitening after every epoch, do they? If we have to choose which epochs implement whitening, this would be easier to do with a callback.

**Comment by sergeyf** [comment]
They do the whitening for every single batch, but they only update the whitening parameters every T batches.

The equation for a dense layer is this:

h_i = f_i (V_i * U_{i-1} * (h_{i-1} − c_i) + d_i )

h_i and h_{i-1} are the outputs of the i-th and (i-1)-th layer, respectively.

f_i is the non-linearity of the i-th layer.

The parameters

**Comment by EderSantana** [comment]
Got it, these parameters that only update every T batches will have to be updated with callbacks, I believe. Did you start implementing it already? Right I got my mind in attention based models right now, but I'd be down to help you out.

**Comment by sergeyf** [comment]
I have not started implementing this! Would be interested in learning how for sure. Do you have an example of how callbacks might work in this case?

Thanks for your help.

**Comment by EderSantana** [comment]
yeah, you will need something like
```python
class UpdateUs(Callback):
    def __init__(self, every_T, U_list):
        self.every_T = every_T
        self.U_list  = U_list
        for u in U_list:
             self.U_updates.append("calculate U updates and append here")
        self.update_U = theano.function([], [], updates=[self.U_list, self.U_updates]) # no inputs or outputs for this theano function, only updates, I'm assuming the whitening won't depend on the input, if it does just pass it in here

    def on_batch_end(self, batch, logs={}):
        if epoch % every_T = 0:
            self.update_U() # Update the U_list
```

I'm assuming you are familiar with Theano and Keras internals, if not, let know which part was not clear.

**Comment by sergeyf** [comment]
Thanks for that snippet. I am only somewhat familiar with internals. 

I see in that snippet how the U_i and c_i are updated using the `on_batch_end`.

Where is h_i = f_i (V_i * U_{i-1} * (h_{i-1} − c_i) + d_i ) actually calculated?

**Comment by EderSantana** [comment]
that is the whitening right?
you will have to do that inside a new Dense layer. Here is how a regular Dense layer works https://github.com/fchollet/keras/blob/master/keras/layers/core.py#L409-L412

You have to do that operation before the dot product I believe.

**Comment by sergeyf** [comment]
Great, thanks. I think those two are enough for me to see if I can get something working.

**Comment by rodrigob** [comment]
@sergeyf how this issue going along ?

**Comment by sergeyf** [comment]
Not very well! My newborn is making me less productive =)
